optuna/samplers/_tpe/sampler.py

```diff
@@ -652,46 +660,29 @@ def _split_complete_trials_multi_objective(
     trials: Sequence[FrozenTrial], study: Study, n_below: int
 ) -> tuple[list[FrozenTrial], list[FrozenTrial]]:
     if n_below == 0:
-        # The type of trials must be `list`, but not `Sequence`.
         return [], list(trials)
 
+    assert 0 < n_below <= len(trials)
     lvals = np.array([trial.values for trial in trials])
     lvals *= np.array([-1.0 if d == StudyDirection.MAXIMIZE else 1.0 for d in study.directions])
-
-    # Solving HSSP for variables number of times is a waste of time.
     nondomination_ranks = _fast_non_domination_rank(lvals, n_below=n_below)
-    assert 0 <= n_below <= len(lvals)
-
-    indices = np.array(range(len(lvals)))
-    indices_below = np.empty(n_below, dtype=int)
-
-    # Nondomination rank-based selection
-    i = 0
-    last_idx = 0
-    while last_idx < n_below and last_idx + sum(nondomination_ranks == i) <= n_below:
-        length = indices[nondomination_ranks == i].shape[0]
-        indices_below[last_idx : last_idx + length] = indices[nondomination_ranks == i]
-        last_idx += length
-        i += 1
-
-    # Hypervolume subset selection problem (HSSP)-based selection
-    subset_size = n_below - last_idx
-    if subset_size > 0:
-        rank_i_lvals = lvals[nondomination_ranks == i]
-        rank_i_indices = indices[nondomination_ranks == i]
-        worst_point = np.max(rank_i_lvals, axis=0)
-        reference_point = np.maximum(1.1 * worst_point, 0.9 * worst_point)
-        reference_point[reference_point == 0] = EPS
-        selected_indices = _solve_hssp(rank_i_lvals, rank_i_indices, subset_size, reference_point)
-        indices_below[last_idx:] = selected_indices
-
-    below_trials = []
-    above_trials = []
-    for index in range(len(trials)):
-        if index in indices_below:
-            below_trials.append(trials[index])
-        else:
-            above_trials.append(trials[index])
+    _, rank_counts = np.unique(nondomination_ranks, return_counts=True)
+    last_rank_before_tiebreak = np.searchsorted(np.cumsum(rank_counts), n_below, side="right") - 1
```

Comment:
if `crc = np.cumsum(rank_counts)`, `crc[i]` is the number of solutions that achieve the non-domination rank of `i` or lower.
`last_rank_before_tiebreak = np.searchsorted(crc, n_below, side="right") - 1` is the integer such that `crc[last_rank_before_tiebreak] <= n_below < crc[last_rank_before_tiebreak]`.
Note that if `last_rank_before_tiebreak == -1`, here we assume that `crc[-1] = 0` instead of the access to the last element of `crc`.